sb_vision/native/apriltag/__init__.py

```diff
@@ -1 +1,98 @@
 """AprilTag native utilities."""
+
+from PIL import Image
+
+from ._apriltag import ffi, lib
+
+
+__all__ = ('AprilTagLibrary',)
+
+
+class AprilTagLibrary:
+    def __init__(self, image_size):
+        """
+        Initialise the AprilTag library.
+
+        This means creating and configuring the detector, which populates a
+        number of tables in memory.
+        """
+        self._image_size = image_size
+
+        self._detector = lib.apriltag_detector_create()
+        """
+        apriltag_detector_t* td,
+        float decimate,
+          default: 1.0, "Decimate input image by this factor"
+        float sigma,
+          default: 0.0, "Apply low-pass blur to input; negative sharpens"
+        int refine_edges,
+          default: 1, "Spend more time trying to align edges of tags"
+        int refine_decode,
+          default: 0, "Spend more time trying to decode tags"
+        int refine_pose
+          default: 0, "Spend more time trying to find the position of the tag"
+        """
+
+        lib.apriltag_init(self._detector, 1.0, 0.0, 1, 0, 0)
+
+        # TODO: keeping this around for the lifetime of the library feels like
+        # it's an optiisation rather than actually required. Given that it makes
+        # the instance non-thread-safe, work out whether it's desirable and/or
+        # there are other reasons the instance isn't thread-safe anyway.
+        self._working_image = lib.image_u8_create_stride(
+            image_size[0],
+            image_size[1],
+            image_size[0],
+        )
+
+    def __del__(self):
+        """Deinitialise the library."""
+
+        if self._detector:
+            lib.apriltag_detector_destroy(self._detector)
+
+        if self._working_image:
+            lib.image_u8_destroy(self._working_image)
+
+    @property
+    def image_size(self):
+        """
+        The image size, a tuple of (width, height), for which the library
+        instance is configured.
+        """
+        return self._image_size
+
+    def detect_tags(self, img: Image):
+        """
+        Run the given image through the apriltags detection library.
+
+        :param img: PIL Luminosity image to be processed
+        :yield: python iterable of apriltag detections; these must be processed
+                and discarded before continuing iteration
+        """
+
+        assert self.image_size == img.size, (
+            "Cannot process images of an incompatible size. Library is "
+            "configured for {}, given image at {}".format(
+                self.image_size,
+                img.size,
+            )
+        )
```

Comment:
Would be more in favour of a `ValueError` in this case rather than using `assert`.

`assert`s can be compiled out when Python is run with `-O`, and this isn't an _internal_ assertion about consistent state: from a user's point of view passing an image of the incorrect size is something that _can_ happen.

Reply:
0571a47